### PR TITLE
CI: Update regression tests for Verilator

### PR DESCRIPTION
**Pull request by marlonjames**
Update regression test suite for Verilator. Add expected test failures and skip some tests because of missing VPI support.

Part of #3194.

Includes alternative to #3235.

Needs #3196 fixed.

### CODE REVIEW

**Comment by marlonjames** [comment]
Do we want Verilator master in the test suite, or some fixed known good version and keep master in the experimental CI run?

**Comment by imphil** [comment]
> Do we want Verilator master in the test suite, or some fixed known good version and keep master in the experimental CI run?

Only use pinned versions in CI to ensure we are actually testing the PR, and not the PR and a new tool version at the same time. So yes, let's keep Verilator master in the extended tests, and use (ideally) the latest release version of Verilator, or if it has to be, a specific development version, in CI.

**Comment by imphil** [comment]
Thanks @marlonjames! 

I am slightly worried that we are hiding problems with Verilator that then bite our users. So if we add test exceptions, we should also make sure that we keep track of these deficiencies in a way that (a) the upstream project knows about them and can eventually get them fixed, and (b) our users know about them as well.

Most of the time we have been using the following approach:
* Have a cocotb issue open for a missing simulator feature, and link to that in the test exception. Ensure that the issue description is understandable to the end user (i.e., not an obscure bug report that then 10 comments down is identified as an simulator issue -- rewrite the issue description if necessary, or create a new issue).
* If it's an upstream issue, mark the issue as "upstream" and link to an upstream bug report. Keep this issue open until it's actually fixed upstream, and verified within cocotb (which then also involves removing the exception).

We are linking to all simulator-related issue in the docs (https://docs.cocotb.org/en/stable/simulator_support.html#sim-verilator-issues), which gives users a reasonably easy way to look for simulator-specific limitations. [I just realized we also show closed issues, I'll do a PR that fixes that next.]

File: cocotb/share/makefiles/simulators/Makefile.verilator [hunk 1]
```diff
@@ -15,11 +15,17 @@ clean::
 
 else
 
-CMD := verilator
+CMD_BIN := verilator
 
-ifeq ($(shell which $(CMD) 2>/dev/null),)
-# Verilator is not in PATH, lets start searching for it
-$(error Cannot find verilator.)
+ifdef VERILATOR_BIN_DIR
```
Comment (imphil) [inline cocotb/share/makefiles/simulators/Makefile.verilator:20]: Is there a specific need for this feature? We require most other simulators to be in PATH.

**Comment by imphil** [comment]
Addition: Looking at https://github.com/cocotb/cocotb/issues/1275 for example I realize that we and I were rather inconsistent in our approach in the past. So let's maybe use this bunch of exceptions to find a consistent approach that we're happy with :smile:

File: cocotb/share/makefiles/simulators/Makefile.verilator (hunk 1 above)
Comment (marlonjames) [reply cocotb/share/makefiles/simulators/Makefile.verilator:20]: Every other simulator makefile supports a `<SIM>_BIN_DIR` environment variable. I also find it useful for my testing setup.

Comment (imphil) [reply cocotb/share/makefiles/simulators/Makefile.verilator:20]: Good with me then, sorry for the noise!

**Comment by ktbarrett** [comment]
@marlonjames Uhhh. This has been here for a while. How is this going?

**Comment by marlonjames** [comment]
It is close. I ran into a few issues with Verilator that I fixed upstream. There are a few left but I am going to make sure they are documented and leave them for now in the interest of getting this done. I should have some time over the next week to wrap it up.

**Comment by toddstrader** [comment]
Awesome.  FYI, `test_package` works now on HEAD of Verilator, but if we want to test against released versions we'll have to wait for the next cut for that test.

**Comment by imphil** [comment]
I'm happy to test against an arbitrary git hash in this PR, and then switch to a released version once that's available.

**Comment by ktbarrett** [comment]
Is there any reason not to test against HEAD?

**Comment by imphil** [comment]
We test against verilator master in the experimental bucket (https://github.com/cocotb/cocotb/blob/cb683a83368af6703a6a326a62d9911603028a66/.github/generate-envs.py#L138-L139). For the CI bucket, we need to have fixed versions.

**Review by imphil — APPROVED** [review]
Great work @marlonjames, thanks a lot for the many issue references and the work you did upstream with Verilator as well!

**Pull request merged**